spectec/src/backend-prose/render.ml

```diff
@@ -24,10 +24,204 @@ let gen_macro env =
 let env config inputs outputs render_latex el prose : env =
   let symbol = Symbol.env el in
   let macro = Macro.env inputs outputs in
-  let hint = Hint.env render_latex el in
-  let env = { config; prose; symbol; macro; hint; } in
+  let env = { config; prose; render_latex; symbol; macro; } in
   env
 
+(* Translation from Al exp to El exp *)
+
+let al_to_el_unop = function
+  | Al.Ast.MinusOp -> Some El.Ast.MinusOp
+  | _ -> None
+
+let al_to_el_binop = function 
+  | Al.Ast.AddOp -> Some El.Ast.AddOp
+  | Al.Ast.SubOp -> Some El.Ast.SubOp
+  | Al.Ast.MulOp -> Some El.Ast.MulOp
+  | Al.Ast.DivOp -> Some El.Ast.DivOp
+  | Al.Ast.ExpOp -> Some El.Ast.ExpOp
+  | _ -> None
+
+let al_to_el_infixop = function
+  | "->" -> Some El.Ast.Arrow
+  | s -> Some (El.Ast.Atom s)
+
+let rec al_to_el_iter iter = match iter with
+  | Al.Ast.Opt -> Some El.Ast.Opt
+  | Al.Ast.List -> Some El.Ast.List
+  | Al.Ast.List1 -> Some El.Ast.List1
+  | Al.Ast.ListN (e, id) ->
+      let ele = al_to_el_exp e in
+      let elid = Option.map (fun id -> id $ no_region) id in
+      (match ele with
+      | Some ele -> Some (El.Ast.ListN (ele, elid))
+      | _ -> None)
+
+and al_to_el_path pl =
+  let fold_path p elp = 
+    let elp' = (match p.it with
+      | Al.Ast.IdxP ei ->
+          let elei = al_to_el_exp ei in
+          (match elei with
+          | Some elei -> Some (El.Ast.IdxP (elp, elei))
+          | _ -> None)
+      | Al.Ast.SliceP (el, eh) ->
+          let elel = al_to_el_exp el in
+          let eleh = al_to_el_exp eh in
+          (match elel, eleh with
+          | Some elel, Some eleh -> Some (El.Ast.SliceP (elp, elel, eleh))
+          | _ -> None)
```

Comment:
The whole translation appears to be inside an option monad. You can simplify the code significantly by using OCaml's `let*` sugar for monadic bind (sort of like `do` in Haskell). To do so, just add the following auxiliary definition to this module:
```
let (let*) = Option.bind
```
and then e.g. the above code can be written without any explicit matches:
```suggestion
          let* elel = al_to_el_exp el in
          let* eleh = al_to_el_exp eh in
          Some (El.Ast.SliceP (elp, elel, eleh))
```
Analogously all the other cases currently matching on an optional result.

Alternatively, it would be even simpler to indicate failure simply by raising an auxiliary exception (especially considering my comment about filter_map below). This can then be caught in the `render_expr` function.

Reply:
Great, this tidies up the code a lot 👍